File: rawDroid/src/main/java/com/anthonymandra/rawdroid/XmpBaseFragment.java
package com.anthonymandra.rawdroid;

import android.content.Context;
import android.content.SharedPreferences;
import android.os.Bundle;
import android.preference.PreferenceManager;
import android.support.v4.app.Fragment;
import android.view.View;
import android.widget.Toast;

import com.anthonymandra.widget.RatingBar;
import com.anthonymandra.widget.XmpLabelGroup;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.Collection;
import java.util.List;

public abstract class XmpBaseFragment extends Fragment implements
		RatingBar.OnRatingSelectionChangedListener,
		XmpLabelGroup.OnLabelSelectionChangedListener,
		KeywordBaseFragment.OnKeywordsSelectedListener
{
	private RatingBar mRatingBar;
	private XmpLabelGroup colorKey;
	private KeywordBaseFragment mKeywordFragment;
	private boolean mPauseListener = false;

	@Override
	public void onViewCreated(View view, Bundle savedInstanceState)
	{
		super.onViewCreated(view, savedInstanceState);
		mRatingBar = (RatingBar) view.findViewById(R.id.ratingBar);
		colorKey = (XmpLabelGroup) view.findViewById(R.id.colorKey);
		mKeywordFragment = (KeywordBaseFragment) getChildFragmentManager().findFragmentById(R.id.keywordFragment);
		attachButtons();
	}

	protected void clear()
	{
//		mPauseListener = true;
		mKeywordFragment.clearSelectedKeywords();
		if (colorKey != null)
			colorKey.clearChecked();
		if (mRatingBar != null)
			mRatingBar.clearChecked();
//		mPauseListener = false;
		onXmpChanged(getXmp());
	}

	protected String[] getSubject()
	{
		Collection<String> selectedKeywords = mKeywordFragment.getSelectedKeywords();
		if (selectedKeywords.size() == 0)
			return null;
		return selectedKeywords.toArray(new String[selectedKeywords.size()]);
	}

	protected Integer[] getRatings()
	{
		List<Integer> ratings = mRatingBar.getCheckedRatings();
		if (ratings.size() == 0)
			return null;
		return ratings.toArray(new Integer[ratings.size()]);
	}

	protected String[] getColorLabels()
	{
		Context c = getContext();
		if (c == null)
			return null;

		SharedPreferences sp = PreferenceManager.getDefaultSharedPreferences(c);
		List<XmpLabelGroup.Labels> checked = colorKey.getCheckedLabels();

		if (checked.size() == 0)
			return null;

		List<String> labels = new ArrayList<>();
		for (XmpLabelGroup.Labels check : checked)
		{
			switch (check)
			{
				case blue:
					labels.add(sp.getString(FullSettingsActivity.KEY_XmpBlue, "Blue"));
					break;
				case red:
					labels.add(sp.getString(FullSettingsActivity.KEY_XmpRed, "Red"));
					break;
				case green:
					labels.add(sp.getString(FullSettingsActivity.KEY_XmpGreen, "Green"));
					break;
				case yellow:
					labels.add(sp.getString(FullSettingsActivity.KEY_XmpYellow, "Yellow"));
					break;
				case purple:
					labels.add(sp.getString(FullSettingsActivity.KEY_XmpPurple, "Purple"));
					break;
			}
		}
		return labels.toArray(new String[labels.size()]);
	}

	protected XmpValues getXmp()
	{
		XmpValues xmp = new XmpValues();
		xmp.label = getColorLabels();
		xmp.rating = getRatings();
		xmp.subject = getSubject();
		return xmp;
	}

	protected abstract void onXmpChanged(XmpValues xmp);

	protected void setColorLabel(String[] labels)
	{
		if (labels == null)
		{
			colorKey.clearChecked();
			return;
		}

		Context c = getContext();
		if (c == null)
			return;

		SharedPreferences sp = PreferenceManager.getDefaultSharedPreferences(c);
		String red = sp.getString(FullSettingsActivity.KEY_XmpRed, "Red");
		String blue = sp.getString(FullSettingsActivity.KEY_XmpBlue, "Blue");
		String green = sp.getString(FullSettingsActivity.KEY_XmpGreen, "Green");
		String yellow = sp.getString(FullSettingsActivity.KEY_XmpYellow, "Yellow");
		String purple = sp.getString(FullSettingsActivity.KEY_XmpPurple, "Purple");

		for (String label : labels)
		{
			if (label.equals(blue))
			{
				colorKey.setChecked(XmpLabelGroup.Labels.blue, true);
			} else if (label.equals(red))
			{
				colorKey.setChecked(XmpLabelGroup.Labels.red, true);
			} else if (label.equals(yellow))
			{
				colorKey.setChecked(XmpLabelGroup.Labels.yellow, true);
			} else if (label.equals(green))
			{
				colorKey.setChecked(XmpLabelGroup.Labels.green, true);
			} else if (label.equals(purple))
			{
				colorKey.setChecked(XmpLabelGroup.Labels.purple, true);
			} else
			{
				Toast.makeText(c, label + " " + getString(R.string.warningInvalidLabel), Toast.LENGTH_LONG).show();
			}
		}
	}

	protected void setXmp(XmpValues xmp)
	{
		setXmp(xmp.rating, xmp.label, xmp.subject);
	}

	protected void setXmp(Integer[] rating, String[] label, String[] subject)
	{
		mPauseListener = true;
		setColorLabel(label);
		setSubject(subject);
		setRating(rating);
		mPauseListener = false;
		onXmpChanged(getXmp());
	}

	/**
	 * Silently set the xmp without firing onXmpChanged
	 * @param xmp xmpValues (rating, label, subject) to init
	 */
	@SuppressWarnings("unused")
	protected void initXmp(XmpValues xmp)
	{
		mPauseListener = true;
		setXmp(xmp);
		mPauseListener = false;
	}

	/**
	 * Silently set the xmp without firing onXmpChanged
	 * @param rating rating
	 * @param label label
	 * @param subject keywords
	 */
	protected void initXmp(Integer[] rating, String[] label, String[] subject)
	{
		mPauseListener = true;
		setColorLabel(label);
		setSubject(subject);
		setRating(rating);
		mPauseListener = false;
	}

	public void reset()
	{
		initXmp(null, null, null);
	}

	protected void setRating(Integer[] ratings)
	{
		mRatingBar.setRating(ratings);
	}

	protected void setExclusive(boolean enable)
	{
		colorKey.setExclusive(enable);
		mRatingBar.setExclusive(enable);
	}

	protected void setAllowUnselected(boolean allow)
	{
		colorKey.setAllowUnselected(allow);
		mRatingBar.setAllowUnselected(allow);
	}

	protected void setSubject(String[] subject)
	{
		if (subject != null)
		{
			mKeywordFragment.setSelectedKeywords(Arrays.asList(subject));
		}
		else
		{
			mKeywordFragment.clearSelectedKeywords();
		}
	}

	private void attachButtons()
	{
		// Ratings
		mRatingBar.setOnRatingSelectionChangedListener(new RatingBar.OnRatingSelectionChangedListener()
		{
			@Override
			public void onRatingSelectionChanged(List<Integer> checked)
			{
				if (!mPauseListener)
					XmpBaseFragment.this.onRatingSelectionChanged(checked);
			}
		});
		colorKey.setOnLabelSelectionChangedListener(new XmpLabelGroup.OnLabelSelectionChangedListener()
		{
			@Override
			public void onLabelSelectionChanged(List<XmpLabelGroup.Labels> checked)
			{
				if (!mPauseListener)
					XmpBaseFragment.this.onLabelSelectionChanged(checked);
			}
		});
		mKeywordFragment.setOnKeywordsSelectedListener(new KeywordEditFragment.OnKeywordsSelectedListener()
		{
			@Override
			public void onKeywordsSelected(Collection<String> selectedKeywords)
			{
				if (!mPauseListener)
					XmpBaseFragment.this.onKeywordsSelected(selectedKeywords);
			}
		});
	}
}

File: app/src/main/java/com/anthonymandra/rawdroid/XmpBaseFragment.kt
package com.anthonymandra.rawdroid

import android.content.SharedPreferences
import android.os.Bundle
import android.preference.PreferenceManager
import android.view.View
import android.widget.Toast
import androidx.fragment.app.Fragment
import com.anthonymandra.rawdroid.data.Label
import com.anthonymandra.rawdroid.data.SubjectEntity
import com.anthonymandra.rawdroid.models.ColorKeys
import com.anthonymandra.rawdroid.settings.MetaSettingsFragment
import com.anthonymandra.widget.RatingBar
import com.anthonymandra.widget.XmpLabelGroup
import java.util.*

abstract class XmpBaseFragment : Fragment(), SharedPreferences.OnSharedPreferenceChangeListener {

    private lateinit var ratingBar: RatingBar
    private lateinit var colorKey: XmpLabelGroup
    private lateinit var keywordFragment: KeywordBaseFragment
    private var mPauseListener = false

	@Suppress("NULLABILITY_MISMATCH_BASED_ON_JAVA_ANNOTATIONS")
	private val colorKeys: ColorKeys by lazy {
        val sp = PreferenceManager.getDefaultSharedPreferences(context)
        sp.registerOnSharedPreferenceChangeListener(this)

		 ColorKeys().also {
            it.blue = sp.getString(MetaSettingsFragment.KEY_XmpBlue, it.blue)
            it.red = sp.getString(MetaSettingsFragment.KEY_XmpRed, it.red)
            it.green = sp.getString(MetaSettingsFragment.KEY_XmpGreen, it.green)
            it.yellow = sp.getString(MetaSettingsFragment.KEY_XmpYellow, it.yellow)
            it.purple = sp.getString(MetaSettingsFragment.KEY_XmpPurple, it.purple)
        }
    }

    protected var subject: List<SubjectEntity>
        get() { return keywordFragment.selectedKeywords.toList() }
        set(value) { keywordFragment.setSelectedKeywords(value) }

    protected var ratings: List<Int>
        get() { return ratingBar.checkedRatings }
        set(value) { ratingBar.setRating(value) }

    protected var colorLabels: List<String>
        get() {
            val checked = colorKey.checked

            val labels = ArrayList<String>()
            checked.forEach { label -> labels.add(colorKeys.customValue(label)) }
            return labels
        }
        set(labels) {
            for (label in labels) {
                when (colorKeys.label(label)) {
                    Label.Blue -> colorKey.setChecked(Label.Blue)
                    Label.Red -> colorKey.setChecked(Label.Red)
                    Label.Green -> colorKey.setChecked(Label.Yellow)
                    Label.Yellow -> colorKey.setChecked(Label.Green)
                    Label.Purple -> colorKey.setChecked(Label.Purple)
                    else -> Toast.makeText(context, label + " " + getString(R.string.warningInvalidLabel), Toast.LENGTH_LONG).show()
                }
            }
        }

    protected var xmp: XmpFilter
        get() {
            return XmpFilter(ratings, colorLabels, subject)
        }
        set(xmp) = setXmp(xmp.rating, xmp.label, xmp.subject)

    @Suppress("NULLABILITY_MISMATCH_BASED_ON_JAVA_ANNOTATIONS")
	 private fun updateColorKey(sp: SharedPreferences, key: String) {
        when (key) {
            MetaSettingsFragment.KEY_XmpBlue ->
                colorKeys.blue = sp.getString(MetaSettingsFragment.KEY_XmpBlue, colorKeys.blue)
            MetaSettingsFragment.KEY_XmpRed ->
                colorKeys.red = sp.getString(MetaSettingsFragment.KEY_XmpRed, colorKeys.red)
            MetaSettingsFragment.KEY_XmpGreen ->
                colorKeys.green = sp.getString(MetaSettingsFragment.KEY_XmpGreen, colorKeys.green)
            MetaSettingsFragment.KEY_XmpYellow ->
                colorKeys.yellow = sp.getString(MetaSettingsFragment.KEY_XmpYellow, colorKeys.yellow)
            MetaSettingsFragment.KEY_XmpPurple ->
                colorKeys.purple = sp.getString(MetaSettingsFragment.KEY_XmpPurple, colorKeys.purple)
        }
    }

    override fun onPause() {
        super.onPause()
        PreferenceManager.getDefaultSharedPreferences(context)
			  .unregisterOnSharedPreferenceChangeListener(this)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        ratingBar = view.findViewById(R.id.ratingBar)
        colorKey = view.findViewById(R.id.colorKey)
        keywordFragment = childFragmentManager.findFragmentById(R.id.keywordFragment) as KeywordBaseFragment

        ratingBar.setOnRatingSelectionChangedListener { checked ->
            if (!mPauseListener)
                this@XmpBaseFragment.onRatingSelectionChanged(checked)
        }

        colorKey.setOnLabelSelectionChangedListener { checked ->
            if (!mPauseListener)
                this@XmpBaseFragment.onLabelSelectionChanged(checked)
        }
        keywordFragment.setOnKeywordsSelectedListener { selectedKeywords ->
            if (!mPauseListener)
                this@XmpBaseFragment.onKeywordsSelected(selectedKeywords)
        }
    }

    protected fun clear() {
        keywordFragment.clearSelectedKeywords()
        colorKey.clearChecked()
        ratingBar.clearChecked()
        onXmpChanged(xmp)
    }

    private fun setXmp(rating: List<Int>, label: List<String>, subject: List<SubjectEntity>) {
        mPauseListener = true
        this.colorLabels = label
        this.subject = subject
        this.ratings = rating
        mPauseListener = false
        onXmpChanged(xmp)
    }

    /**
     * Silently set the xmp without firing onXmpChanged
     * @param rating rating
     * @param label label
     * @param subject keywords
     */
    protected fun initXmp(rating: List<Int> = Collections.emptyList(),
                          label: List<String> = Collections.emptyList(),
                          subject: List<SubjectEntity> = Collections.emptyList()) {
        mPauseListener = true
        this.colorLabels = label
        this.subject = subject
        this.ratings = rating
        mPauseListener = false
    }

    fun reset() {
        initXmp()
    }

    protected fun isSingleSelection(enable: Boolean) {
        colorKey.isSingleSelection = enable
        ratingBar.isSingleSelection = enable
    }

//    protected fun setAllowUnselected(allow: Boolean) {
//        colorKey.setAllowUnselected(allow)
//        ratingBar.setAllowUnselected(allow)
//    }

    protected abstract fun onXmpChanged(xmp: XmpFilter)
    abstract fun onKeywordsSelected(selectedKeywords: Collection<SubjectEntity>)
    abstract fun onRatingSelectionChanged(checked: List<Int>)
    abstract fun onLabelSelectionChanged(checked: List<Label>)

    override fun onSharedPreferenceChanged(sharedPreferences: SharedPreferences, key: String) {
        updateColorKey(sharedPreferences, key)
    }
}
